File: algorithms/16_kotlin/sorting/InsertionSort.kt
fun insertionSort(arr: IntArray): IntArray {
    val a = arr.copyOf()
    for (i in 1 until a.size) {
        val key = a[i]
        var j = i - 1
        while (j >= 0 && a[j] > key) {
            a[j + 1] = a[j]
            j--
        }
        a[j + 1] = key
    }
    return a
}

fun main() {
    val data = intArrayOf(33, 7, 91, 12, 5, 5, 78, 2, 44, 19)
    val sorted = insertionSort(data)
    check(sorted.contentEquals(data.sortedArray()))
    println("[Kotlin InsertionSort] Insertion sort verified: ${sorted.joinToString(", ")}")
}

File: algorithms/11_golang/sorting/insertion_sort.go
package main

import "fmt"

func insertionSort(a []int) {
	for i := 1; i < len(a); i++ {
		key := a[i]
		j := i - 1
		for j >= 0 && a[j] > key {
			a[j+1] = a[j]
			j--
		}
		a[j+1] = key
	}
}

func main() {
	data := []int{33, 7, 91, 12, 5, 5, 78, 2, 44, 19}
	insertionSort(data)
	for i := 1; i < len(data); i++ {
		if data[i-1] > data[i] {
			panic("not sorted")
		}
	}
	fmt.Println("[Go InsertionSort] Insertion sort verified:", data)
}
